Reject password inputs looking like raw bcrypt

Alkarex opened the pull request:
If a password arrives in the authorization header as a string looking like bcrypt (according to standard Modular Crypt Format) instead of normal-looking plain-text, then reject the authentication.
This is to avoid being able to pass the crypted version of the password, in case it was leaked.

Alkarex committed:
Reject password inputs looking like raw bcrypt
If a password arrives in the authorization header as a string looking like bcrypt (according to standard Modular Crypt Format) instead of normal-looking plain-text, then reject the authentication.
This is to avoid being able to pass the crypted version of the password, in case it was leaked.

quackzar commented on test.sh:
```diff
@@ -44,3 +52,16 @@ if [ "$test" != "" ]; then
+
+test=$(
+	cat <<'EOF' | node ./index.js http-basic-auth --realm='"node-red"' --username='"test"' --password='"$2y$10$5TSZDldoJ7MxDZdtK/SG2O3cwORqLDhHabYlKX9OsM.W/Z/oLwKW6"'
+{"req":{"headers":{"authorization":"Basic dGVzdDokMnkkMTAkNVRTWkRsZG9KN014RFpkdEsvU0cyTzNjd09ScUxEaEhhYllsS1g5T3NNLlcvWi9vTHdLVzY="}}}
```
It might be less opaque if the test input isn't hardcoded in base64 but just calls the `base64` command as a subshell or using a variable.

Alkarex replied:
Good idea. Here you are https://github.com/alexandrainst/node-red-http-basic-auth/pull/9/commits/c6842f50b6d144d81ef8f060760c10b981718c39

quackzar commented on nodes/http-auth.js:
```diff
@@ -13,21 +13,32 @@ try {
+ * Is the password matching the Modular Crypt Format of bcrypt
+ */
+function isBcryptMcf(hash) {
+	return (typeof hash === 'string') && hash.match(/^\$2[abxy]?\$/);
```
While unlikely, we now reject some passwords if they match that regex. While unlikely, it might be an issue for some esoteric password generators.

Alkarex replied:
Yes indeed. There is the same behaviour in e.g. Apache htpasswd approach, which I try to be compatible with

quackzar replied:
That would probably be the better solution then

Alkarex committed:
Use explicit base64 in test examples
https://github.com/alexandrainst/node-red-http-basic-auth/pull/9#discussion_r1387642267
